packages/protocol/test/L1/TaikoL1.integration.test.ts

```diff
@@ -0,0 +1,160 @@
+            (
+                await l2Provider.listAccounts()
+            )[0]
+        );
+
+        taikoL2 = await deployTaikoL2(l2Signer);
+
+        const genesisHash = taikoL2.deployTransaction.blockHash as string;
+
+        taikoL1 = await deployTaikoL1(genesisHash);
+    });
+
+    describe("isCommitValid()", async function () {
+        it("should not be valid", async function () {
+            const block = await l2Provider.getBlock("latest");
+            const commit = generateCommitHash(block);
+
+            const isCommitValid = await taikoL1.isCommitValid(
+                1,
+                1,
+                commit.hash
+            );
+
+            expect(isCommitValid).to.be.eq(false);
+        });
+    });
+
+    describe("getProposedBlock()", function () {
+        it("proposed block does not exist", async function () {
+            const block = await taikoL1.getProposedBlock(123);
+            expect(block[0]).to.be.eq(ethers.constants.HashZero);
+            expect(block[1]).to.be.eq(ethers.constants.AddressZero);
+            expect(block[2]).to.be.eq(BigNumber.from(0));
+        });
+    });
+    describe("commitBlock() -> proposeBlock() integration", async function () {
+        it("should revert with invalid meta", async function () {
+            const block = await l2Provider.getBlock("latest");
+            const { tx, commit } = await commitBlock(taikoL1, block);
+
+            await expect(
+                proposeBlock(
+                    taikoL1,
+                    block,
+                    commit.txListHash,
+                    tx.blockNumber as number,
+                    1,
+                    block.gasLimit
+                )
+            ).to.be.revertedWith("L1:placeholder");
+        });
+
+        it("should revert with invalid gasLimit", async function () {
+            const block = await l2Provider.getBlock("latest");
+            const { tx, commit } = await commitBlock(taikoL1, block);
+
+            // blockMetadata is inputs[0], txListBytes = inputs[1]
+            const config = await taikoL1.getConfig();
+            const gasLimit = config[7];
+            await proposeBlock(
+                taikoL1,
+                block,
+                commit.txListHash,
+                tx.blockNumber as number,
+                0,
+                block.gasLimit
+            );
+
+            await expect(
+                proposeBlock(
+                    taikoL1,
+                    block,
+                    commit.txListHash,
+                    tx.blockNumber as number,
+                    0,
+                    gasLimit.add(1)
+                )
+            ).to.be.revertedWith("L1:gasLimit");
+        });
+
+        it("should revert with invalid extraData", async function () {
+            const block = await l2Provider.getBlock("latest");
+            const { tx, commit } = await commitBlock(taikoL1, block);
+
+            const meta: BlockMetadata = {
```

Comment:
Maybe we can define a function `metaForBlock(block)`, then

```
let meta = metaForBlock(block) // place holder fields all set to 0
meta.extraData = ethers.utils.hexlify(ethers.utils.randomBytes(33))
...
```

Reply:
If we have the `metaForBlock(block)` function, I'd suggest we delete the function `proposeBlock` and use the following pattern instead:

```
const inputs = buildProposeBlockInputs(block, meta);
taikoL1.proposeBlock(inputs)
```